### DISCUSSION TITLE
[RFC] Data Request Format version 3

### DISCUSSION THREAD

**Discussion by beauremus** [fermi-ad/controls-issues#31] [Request for Comments (RFCs)]
### Basic example

`<device-name>[device-qualifiers][@event-specifier]`

The `device-name` is required, indicated by the `<>`. The `[]` groups are optional and grouped with their designated separator.

### Motivation

The Data Request Format (DRF) has proven its worth, especially with the growing demand for programmatic access to accelerator data. DRF allows users to succinctly express a complex data request with a simple structured string.

With the addition of EPICS to our accelerator complex, DRF version 2 (DRF2) needs to be redesigned to be more general, supporting not only EPICS but also future control systems.

DRF version 3 (DRF3) would need to support device names from both ACNET and EPICS. Ideally, the parsing strategy allows for arbitrary device names.

DRF2 supports periodic and event requests, but we are aware of requests for a more sophisticated arming and triggering data request strategy that would be nice to support with DRF3.

Out of necessity, DRF2 was extended in the Data Pool Manager (DPM) to support various data sources, including live (default), FTP (for data greater than 15Hz), LOGGER (for historical data), SRFILE (for saving and restoring save states), and REDIR (allows redirecting data requests to an arbitrary data source). We should take this idea for DRF3, generalize it, and expand upon it.

_Note:_ All of these features require the rest of the control system to support these ideas. This is simply a way to express a complex request to the rest of the accelerator complex.

### Detailed design

TBD

### Drawbacks

I think it's reasonable to ask, "Why use a string?" when the motivation is programmatic access to accelerator data. Using a tool that allows programmers to use native data structures could be more appealing.

Using a string structure requires Controls to maintain the DRF specification and the parser implementation.

In my experience, users need expert assistance in assembling the desired request despite having a specification document for DRF2. I'm not sure other strategies alleviate this, but we should consider the complexity of the learning curve and attempt to offer a comprehensive set of example requests.

If we introduce a more complex arm and trigger system, users will expect it to be supported, requiring work from front-end programmers.

Ideally, this is backward compatible with DRF2, but introducing a new control system may be the time to make any necessary breaking changes. The costs of migration are the costs of implementation mentioned above.

### Alternatives

I couldn't find other examples of structured strings, so please offer alternatives if you can find examples.

gRPC, GraphQL, and JSON would all be good alternatives to the structured string approach. These strategies benefit from the ease of representing the data natively in many different languages. These solutions would result in more verbose requests as they require key-value pairs.

### Adoption strategy

It still needs to be made clear if this is a breaking change. Ideally, it is not.

DPM owns the current implementation of the DRF parser. The future DRF3 parser should be implemented as an independent service that allows all applications and services to use its parsing results.

### How we teach this

The core concepts of DRF are the same, so any existing references aren't required to change.

Any existing references to the detailed use of DRF that are modified by this spec would need to be updated.

I would like to see the DRF structure, the idea of a device, event, and data source be used in all applications and services. This standardization provides a more consistent user experience and sets the expectation for how requests are made and what features are available from a data request.

### Unresolved questions

- Should the components of a DRF be broken into separate specs?
  - For example, the data logger only uses the `event-specifier`.
- What does the arm-trigger part of the `event-specifier` look like?
- Can we define a standard set of `device-qualifiers`, or do we leave it up to the user to know what qualifiers a device supports?
- Should this specification prescribe a strategy for errors and reporting?
  - Should services give a best effort to match a request or fail quickly with an unsupported feature in the request?
- What data sources should be specified?
- Can we create something useful before ACORN has made changes, or should this wait until significant ACORN changes are in place?

## Edits:

- Changed example optional groups to appropriately signify dependent and independent _groups_.
- Remove data-sources. This functionality should be defined by the service DRF is passed to.

**Comment by beauremus** [comment]
After recent conversations looking at supporting ACNET and PVAccess requests, we found that this structure works for both use cases.

Original DRF2 BNF:
`request = device [ "." property ] [ range ] [ "." field ] [ "@" event ]`

Proposed DRF3 BNF:
`request = device [ "." field ]* [ range ] [ "@" event ]`

The "*" after the field indicates that there can be zero or more fields.

This strategy makes translating the existing ACNET model to the DRF3 strategy straightforward.

**Reply by ebmilton** [reply]
Adam is using your RFC as an example. I would like to wholly disagree without reading the RFC, as an example of how users will wholly disagree without reading the RFC.
